mater/lib/src/stores/file.rs

```diff
@@ -0,0 +1,296 @@
+use std::{io::SeekFrom, path::Path};
+
+use indexmap::IndexMap;
+use ipld_core::cid::Cid;
+use sha2::{Digest, Sha256};
+use tokio::{
+    fs::File,
+    io::{AsyncSeekExt, AsyncWriteExt},
+    sync::{Mutex, RwLock},
+};
+
+use crate::{
+    multicodec::SHA_256_CODE,
+    v1::{self, read_block, write_block},
+    v2::{self},
+    CarV1Header, CarV2Header, Characteristics, Error, Index, IndexEntry, MultihashIndexSorted,
+    SingleWidthIndex,
+};
+
+/// Implements a blockstore that stores blocks in a CARv2 format. Blocks put
+/// into the blockstore can be read back once they are successfully written. The
+/// blocks are written immediately, while the index is stored in memory and
+/// updated incrementally.
+///
+/// The blockstore should be closed once the putting blocks is finished. Upon
+/// closing the blockstore, the index is written out to underlying file.
+pub struct FileBlockstore {
+    // Inner store
+    inner: Mutex<FileBlockstoreInner>,
+    // Index of blocks that will be appended to the file at the finalization
+    index: RwLock<IndexMap<Cid, u64>>,
+}
+
+/// Inner file store. Encapsulating state that is locked and used together.
+struct FileBlockstoreInner {
```

Comment:
Could you remind me how it's gonna be used in the Storage Provider Server?

Reply:
FileBlockstore will be used by the client. Basically the client requests blocks from the providers and then stores the received blocks to the FileBlockstore.

Reply:
Why does the client uses FileBlockstore to store the files?
I mean, when you download the file, what's the point of storing it in the blockstore?

Reply:
The content is transferred in cid-addressed blocks. [This](https://docs.ipfs.tech/assets/img/diagram-of-the-want-have-want-block-process.6ef862a2.png) image shows nicely how the protocol works. The client requests and receives blocks from the car archive. Then when it gets those blocks back it stores them in the blockstore.

You could imagine it something like, the blockstore is an intermediary state that is used to store the car file received as a stream of blocks. The blockstore is finalized (the index and correct header is written to disk) when the client sees that there are no blocks needed to be requested anymore. When the blockstore is finalized, client is left with the correct car file on disk.

Reply:
Ok, so when client wants to retrieve a file, it needs to ask for each of the blocks in the CARv2 archive and then reconstruct it on their own. How does the client know which blocks to ask for?

Reply:
1. The client gets the payload (root) cid of the car file from the user. That means the client knows which is the first block that needs to be fetched. If the received block represents a node in the DAG (is not a leaf). Then it decodes the content. By doing that it gets the next cids it has to request. The client is doing this until there are some blocks left to request.

2. The easiest and fastest client implementation is to request the blocks in sequence. That way the client knows that the blocks received are in sequence and they can be appended to the file. The improvement would be to implement some kindof linking system that would hold the state of the links.

3. The reconstruction is done on the fly by the blockstore in this PR. So to quickly sum up. The client requests the blocks sequentially. The received blocks are added to the FileBlockstore and as a side effect the car is being build on the disk. After all blocks are requested and received, the blockstore is finalized by the client.